File: src-tauri/src/state.rs
//! Managed state carrying the dsh child process id, and process-group teardown.

use std::sync::Mutex;
use std::time::Duration;

pub struct DshState {
    /// The child process id; also the process-group id (see `process_group(0)`).
    /// Cleared after a kill so teardown is idempotent.
    pid: Mutex<Option<u32>>,
}

impl DshState {
    pub fn new(pid: u32) -> Self {
        Self {
            pid: Mutex::new(Some(pid)),
        }
    }

    pub fn pid(&self) -> Option<u32> {
        *self.pid.lock().unwrap()
    }

    /// Terminate the dsh process group (SIGTERM, then SIGKILL after a grace
    /// period). The pid is cleared so repeated calls become a no-op.
    pub fn kill(&self) {
        let mut guard = self.pid.lock().unwrap();
        let Some(pid) = guard.take() else {
            return;
        };
        unsafe {
            libc::kill(-(pid as i32), libc::SIGTERM);
        }
        std::thread::sleep(Duration::from_millis(1500));
        unsafe {
            libc::kill(-(pid as i32), libc::SIGKILL);
        }
    }
}

File: app/src/state.js
'use strict'

// Managed state carrying the dsh child process, process-group teardown, and shutdown intent
// (Electron port of state.rs). The child is spawned detached (its own process group), so the
// group id equals the child pid.

class DshState {
  constructor(child) {
    this._child = child
    this._pid = child && child.pid != null ? child.pid : (typeof child === 'number' ? child : null)
    this.shuttingDown = false
  }

  pid() {
    return this._pid
  }

  /** Take and clear the pid, making teardown idempotent. */
  takePid() {
    const pid = this._pid
    this._pid = null
    return pid
  }

  /** Send a signal to the whole dsh process group. */
  signal(sig) {
    const pid = this._pid
    if (pid === null || pid === undefined) return
    try {
      process.kill(-pid, sig)
    } catch {
      // Process group already gone; ignore.
    }
  }

  /**
   * Graceful teardown: SIGTERM now, then SIGKILL after a grace period.
   * 观察子进程退出以提前收尾(不再傻等 graceMs);`done` 在 teardown 后调用
   * (或没有 pid 时立即调用)。置位 shuttingDown,用于区分主动退出与崩溃。
   */
  killGracefully(graceMs, done) {
    this.shuttingDown = true
    const pid = this.takePid()
    if (pid === null || pid === undefined) {
      if (done) done()
      return
    }
    let finished = false
    let timer
    const finish = () => {
      if (finished) return
      finished = true
      clearTimeout(timer)
      if (done) done()
    }
    const child = this._child
    if (child && typeof child.once === 'function') child.once('exit', finish)
    try {
      process.kill(-pid, 'SIGTERM')
    } catch {}
    timer = setTimeout(() => {
      try {
        process.kill(-pid, 'SIGKILL')
      } catch {}
      finish()
    }, graceMs)
  }
}

module.exports = { DshState }
